dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java

```diff
@@ -0,0 +1,370 @@
+
+    @Autowired
+    private TrackedEntityAttributeService trackedEntityAttributeService;
+
+    @Autowired
+    private ProgramService programService;
+
+    @Autowired
+    private OrganisationUnitService organisationUnitService;
+
+    @Autowired
+    private CategoryService categoryService;
+
+    @Autowired
+    private DataElementService dataElementService;
+
+	@Override
+	public boolean accept( IncomingSms sms ) {
+        if ( sms == null || !SmsUtils.isBase64(sms) )
+        {
+            return false;
+        }
+
+        return handlesType(getHeader(sms).getType());
+	}
+
+	@Override
+	public void receive( IncomingSms sms ) {
+		SMSSubmissionHeader header = null;
+		
+		try {
+			SMSSubmissionReader reader = new SMSSubmissionReader();
+			header = reader.readHeader(SmsUtils.getBytes(sms));
+			SMSMetadata meta = getMetadata(header.getLastSyncDate());
+			SMSSubmission subm = reader.readSubmission(SmsUtils.getBytes(sms), meta);
+						
+			//TODO: Can be removed - debugging line to check SMS submissions
+			Gson gson = new GsonBuilder().setPrettyPrinting().create();
+			Log.info("New received SMS submission decoded as: " + gson.toJson(subm));
+			
+			checkUserAndOrgUnit(subm);
+			SMSResponse resp = postProcess(sms, subm);
+			
+			Log.info("SMS Response: " + resp.toString());
+	        update(sms, SmsMessageStatus.PROCESSED, true);
+	        sendSMSResponse(resp, sms.getOriginator(), header.getSubmissionID());
+		} catch ( SMSProcessingException e ) {
+			Log.error(e.getMessage());
+			update(sms, SmsMessageStatus.FAILED, true);
+			sendSMSResponse(e.getResp(), sms.getOriginator(), header.getSubmissionID());
+		} catch ( Exception e ) {
+			//Exceptions caught here will come from reading the header and submission
+			//itself. We may not even have a submission ID to respond to the app with.
+			//TODO: We can handle exceptions in read submission assuming we have a
+			// 		valid header. We can't do anything about CRC checks though.
+			Log.error(e.getMessage());
+			update(sms, SmsMessageStatus.FAILED, true);
+			if (header != null) {
+				sendSMSResponse(SMSResponse.UNKNOWN_ERROR, sms.getOriginator(), header.getSubmissionID());
+			}
+		}
+	}
+
+	private void checkUserAndOrgUnit( SMSSubmission subm )
+	{
+		String ouid = subm.getOrgUnit();
+		String userid = subm.getUserID();
+        OrganisationUnit ou = organisationUnitService.getOrganisationUnit(ouid);
+        User user = userService.getUser(userid);
+
+        if (ou == null) {
+        	throw new SMSProcessingException(SMSResponse.INVALID_ORGUNIT.set(ouid));
+        } else if (user == null) {
+        	throw new SMSProcessingException(SMSResponse.INVALID_USER.set(userid));
+        } else if (!user.getOrganisationUnits().contains(ou)) {
+        	throw new SMSProcessingException(SMSResponse.USER_NOTIN_OU.set(userid, ouid));
+        }
+	}
+	
+	private SMSSubmissionHeader getHeader( IncomingSms sms ) {
+		byte[] smsBytes = SmsUtils.getBytes(sms);
```

Comment:
In case of any error in decoding SmsUtils.getBytes will return null

Reply:
Yeah I can send an SMS error back in getHeader as well, the trick is that I won't have a message ID to tell the app which message failed. I'm leaning towards just sending back a message with a '?' for the message ID though so at least the app has something.